models/nft/ethereum/nft_ethereum_wallet_pnl.sql

```diff
@@ -0,0 +1,149 @@
+    AND buyer != seller 
+    AND trade_type = 'Single Item Trade'
+    AND amount_original IS NOT NULL 
+    AND block_time < NOW() - Interval '5 Minutes'
+    GROUP BY 1, 2, 3, 6
+    {% endif %}
+    -- incremental run 
+    {% if is_incremental() %}
+        seller as wallet, 
+        nft_contract_address, 
+        'Sells' as trade_type,
+        SUM(amount_original) as eth_amount, 
+        COUNT(*) as trades, 
+        NOW() - Interval '5 Minutes' as last_updated
+    FROM 
+    {{ ref('nft_trades') }}
+    WHERE block_time >= (SELECT MAX(last_updated) - Interval '5 Minutes' FROM {{this}})
+    AND currency_symbol IN ('ETH', 'WETH')
+    AND blockchain = 'ethereum'
+    AND buyer != seller 
+    AND trade_type = 'Single Item Trade'
+    AND amount_original IS NOT NULL 
+    AND block_time < NOW() - Interval '5 Minutes'
+    GROUP BY 1, 2, 3, 6
+
+    UNION ALL 
+
+    SELECT 
+        buyer as wallet, 
+        nft_contract_address, 
+        'Buys' as trade_type,
+        -1 * SUM(amount_original) as eth_amount,
+        COUNT(*) as trades, 
+        NOW() - Interval '5 Minutes' as last_updated
+    FROM 
+    {{ ref('nft_trades') }}
+    WHERE block_time >= (SELECT MAX(last_updated) FROM {{this}})
+    AND currency_symbol IN ('ETH', 'WETH')
+    AND blockchain = 'ethereum'
+    AND buyer != seller 
+    AND trade_type = 'Single Item Trade'
+    AND amount_original IS NOT NULL 
+    AND block_time < NOW() - Interval '5 Minutes'
+    GROUP BY 1, 2, 3, 6
+    {% endif %}
+), 
+
+combined as (
+    SELECT 
+     {% if not is_incremental() %}
+        wallet, 
+        nft_contract_address, 
+        last_updated,
+        COALESCE(SUM(CASE 
+            WHEN trade_type = 'Buys'
+            THEN ABS(eth_amount) 
+            ELSE 0 
+        END), 0) as eth_spent, 
+        COALESCE(SUM(CASE 
+            WHEN trade_type = 'Sells'
+            THEN eth_amount
+            ELSE 0 
+        END), 0) as eth_received, 
+        SUM(eth_amount) as pnl, 
+        SUM(trades) as trades 
+    FROM 
+    trades 
+    GROUP BY 1, 2, 3
+    {% endif %}
+    {% if is_incremental() %}
+    SELECT 
+        prev.wallet, 
+        prev.nft_contract_address,
+        t.last_updated,
+        COALESCE(SUM(CASE 
+            WHEN t.trade_type = 'Buys'
+            THEN ABS(t.eth_amount) 
+            ELSE 0 
+        END), 0) + prev.eth_spent as eth_spent,
+        COALESCE(SUM(CASE 
+            WHEN t.trade_type = 'Sells'
+            THEN t.eth_amount
+            ELSE 0 
+        END), 0) + prev.eth_received as eth_received,
+        COALESCE(SUM(t.eth_amount), 0) +  prev.pnl as pnl, 
+        COALESCE(SUM(t.trades), 0) + prev.trades as trades 
+    FROM 
+    {{this}} prev 
+    LEFT JOIN 
+    trades t 
+        ON prev.wallet = t.wallet 
+        AND prev.nft_contract_address = t.nft_contract_address
+    GROUP BY 1, 2, 3
+    {% endif %}
+)
+
+SELECT 
+    ROW_NUMBER() OVER (PARTITION BY nft_contract_address ORDER BY pnl DESC) as index, 
```

Comment:
yess, so this spell returns the pnl of the wallets for a specific nft_contract_address so yeah it can serve and is enough as the unique key. 

Also, just curious, wouldn't merge cause duplicates here? if we're using nft_contract_address & wallet as the unique keys then what happens seeing as the pnl & rank would possible change?

Reply:
hmmm, the index would probably be problematic. I think we can leave that out of the materialisation tbh. Seems like it will add too much complexity, looking at how the query is set up.